 ## Importing the necessary Python Libraries for the Quantitative Analysis of the Stock Market.

In [1]:
#import pandas and plotly
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import plotly.io as pio
pio.templates.default = "plotly_white"

## Importing the Dataset for the Quantitative Analysis of the Stock Market.

In [18]:
stocks_data = pd.read_csv('stocks.csv')
stocks_data.info()
stocks_data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 248 entries, 0 to 247
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Ticker     248 non-null    object 
 1   Date       248 non-null    object 
 2   Open       248 non-null    float64
 3   High       248 non-null    float64
 4   Low        248 non-null    float64
 5   Close      248 non-null    float64
 6   Adj Close  248 non-null    float64
 7   Volume     248 non-null    int64  
dtypes: float64(5), int64(1), object(2)
memory usage: 15.6+ KB


Ticker        Date        Open        High         Low       Close  \
0     AAPL  2023-02-07  150.639999  155.229996  150.639999  154.649994   
1     AAPL  2023-02-08  153.880005  154.580002  151.169998  151.919998   
2     AAPL  2023-02-09  153.779999  154.330002  150.419998  150.869995   
3     AAPL  2023-02-10  149.460007  151.339996  149.220001  151.009995   
4     AAPL  2023-02-13  150.949997  154.259995  150.919998  153.850006   
..     ...         ...         ...         ...         ...         ...   
243   GOOG  2023-05-01  107.720001  108.680000  107.500000  107.709999   
244   GOOG  2023-05-02  107.660004  107.730003  104.500000  105.980003   
245   GOOG  2023-05-03  106.220001  108.129997  105.620003  106.120003   
246   GOOG  2023-05-04  106.160004  106.300003  104.699997  105.209999   
247   GOOG  2023-05-05  105.320000  106.440002  104.738998  106.214996   

      Adj Close    Volume  
0    154.414230  83322600  
1    151.688400  64120100  
2    150.639999  56007100  
3    151.009995  57450700  
4    153.850006  62199000  
..          ...       ...  
243  107.709999  20926300  
244  105.980003  20343100  
245  106.120003  17116300  
246  105.209999  19780600  
247  106.214996  20705300  

[248 rows x 8 columns]

## Using Statistical Concepts to perform the Quantitative Analysis of the Stock Market
1. Descriptive Statistics.
2. Time Series Analysis.
3. Volatility Analysis.
4. Correlation Analysis.
5. Comparative Analysis.
6. Risk vs Return Analysis.

### 1. Descriptive Statistics
- Summary statistics (mean, median, standard deviation, etc) for each stock closing prices.

In [26]:
descriptive_stats = stocks_data.groupby('Ticker')['Close'].describe() 
descriptive_stats

count        mean        std         min         25%         50%  \
Ticker                                                                     
AAPL     62.0  158.240645   7.360485  145.309998  152.077499  158.055000   
GOOG     62.0  100.631532   6.279464   89.349998   94.702501  102.759998   
MSFT     62.0  275.039839  17.676231  246.270004  258.742500  275.810013   
NFLX     62.0  327.614677  18.554419  292.760010  315.672493  325.600006   

               75%         max  
Ticker                          
AAPL    165.162506  173.570007  
GOOG    105.962503  109.459999  
MSFT    287.217506  310.649994  
NFLX    338.899994  366.829987

###### Insight from the result for each stock:

###### AAPL (Apple Inc.)
- Count:                 62.0 (The number of observations or trading days included in the dataset for AAPL)
- Mean:                  158.24 (The average closing price)
- Standard Deviation:    7.36 (Measures the amount of variation or dispersion of closing prices)
- Minimum:               145.31 (The lowest closing price in the dataset)
- 25th Percentile:       152.08 (25% of the closing prices are below this value)
- Median (50%):          158.06 (The middle value of the closing prices)
- 75th Percentile:       165.16 (75% of the closing prices are below this value)
- Maximum:               173.57 (The highest closing price in the dataset)

###### GOOG (Alphabet Inc.)
- Similar statistics as AAPL, but for GOOG. The mean closing price is 100.63, with a standard deviation of 6.28, indicating less variability in closing prices compared to AAPL.

###### MSFT (Microsoft Corporation)
- The dataset includes the same number of observations for MSFT. It has a higher mean closing price of 275.04 and a higher standard deviation of 17.68, suggesting greater price variability than AAPL and GOOG.

###### NFLX (Netflix Inc.)
#NFLX shows the highest mean closing price (327.61) among these stocks and the highest standard deviation (18.55), indicating the most significant price fluctuation.

### 2. Time Series Analysis
- To examine trends and patterns over time, focusing on the closing prices.

In [28]:
stocks_data['Date'] = pd.to_datetime(stocks_data['Date'])
pivot_data = stocks_data.pivot(index = 'Date', columns = 'Ticker', values = 'Close')

In [32]:
# creating subplot
fig = make_subplots(rows = 1, cols = 1)

In [34]:
# adding traces for each stock ticker
for column in pivot_data.columns:
    fig.add_trace(go.Scatter(x = pivot_data.index, y = pivot_data[column], name = column), row = 1, col = 1)

In [36]:
# updating the layout
fig.update_layout(
    title_text = 'Time Series of Closing Prices',
    xaxis_title = 'Date',
    yaxis_title = 'Closing Price',
    legend_title = 'Ticker',
    showlegend = True)

# show plot
fig.show()

###### Insight from the Time Series Analysis - Key observations from the closing prices for each stock (AAPL, GOOG, MSFT, NFLX):

###### Trend: 
- Each stock shows its unique trend over time. For instance, AAPL and MSFT exhibit a general upward trend in this period.
###### Volatility: 
- There is noticeable volatility in the stock prices. For example, NFLX shows more pronounced fluctuations compared to others.
###### Comparative Performance: 
- When comparing the stocks, MSFT and NFLX generally trade at higher price levels than AAPL and GOOG in this dataset.

### 3. Volatility Analysis.
- To calculate and compare the volatility (standard deviation) of the closing prices for each stock. This will give us an insight into how much the stock prices fluctuated over the period.

In [38]:
volatility = pivot_data.std().sort_values(ascending = False)

In [39]:
fig = px.bar(volatility,
            x = volatility.index,
            y = volatility.values,
            labels = {'y': 'Standard Deviation', 'x': 'Ticker'},
            title = 'Volatility of Closing Prices(Standard Deviation)')
fig.show()

###### Insight from the Volatility Analysis of Closing Prices (Standard Deviation) 

- The bar chart and the accompanying data show the volatility (measured as standard deviation) of the closing prices for each stock. Here’s how they rank in terms of volatility:
- NFLX: Highest volatility with a standard deviation of approximately 18.55.
- MSFT: Next highest, with a standard deviation of around 17.68.
- AAPL: Lower volatility compared to NFLX and MSFT, with a standard deviation of about 7.36.
- GOOG: The least volatile in this set, with a standard deviation of approximately 6.28.
- It indicates that NFLX and MSFT stocks were more prone to price fluctuations during this period compared to AAPL and GOOG.

### 4. Correlation Analysis.
- To understand how the stock prices of these companies are related to each other.

In [40]:
correlation_matrix = pivot_data.corr()

In [41]:
# creating Heatmap
fig = go.Figure(data = go.Heatmap(
                    z = correlation_matrix,
                    x = correlation_matrix.columns,
                    y = correlation_matrix.columns,
                    colorscale = 'blues',
                    colorbar = dict(title = 'Correlation'),
                    ))

In [43]:
# updating layout
fig.update_layout(
    title = 'Correlation Matrix of Closing Prices',
    xaxis_title = 'Ticker',
    yaxis_title = 'Ticker'
)

fig.show()

###### Insight from the Correlation Analysis of Closing Prices

- The heatmap displays the correlation matrix of the closing prices of the four stocks (AAPL, GOOG, MSFT, NFLX). Here’s what the correlation coefficients suggest:

- Values close to +1 indicate a strong positive correlation, meaning that as one stock’s price increases, the other tends to increase as well.
- Values close to -1 indicate a strong negative correlation, where one stock’s price increase corresponds to a decrease in the other.
- Values around 0 indicate a lack of correlation.

- From the heatmap, we can observe that there are varying degrees of positive correlations between the stock prices, with some pairs showing stronger correlations than others. For instance, AAPL and MSFT seem to have a relatively higher positive correlation.


### 5. Comparative Analysis.
- To compare the performance of different stocks based on their returns over the period. Calculate the percentage change in closing prices from the start to the end of the period for each stock.

In [44]:
# calculating the percentage change in closing prices
percentage_change = ((pivot_data.iloc[-1] - pivot_data.iloc[0]) / pivot_data.iloc[0]*100)

In [46]:
# creating a column chart
fig = px.bar(volatility,
            x = percentage_change.index,
            y = percentage_change.values,
            labels = {'y': 'Percentage Change (%)', 'x': 'Ticker'},
            title = 'Percentage Change in Closing Prices')

fig.show()

###### Insight from the Comparative Analysis - Percentage Change in Closing Prices:

- The bar chart and the accompanying data show the percentage change in the closing prices of the stocks from the start to the end of the observed period:

- MSFT: The highest positive change of approximately 16.10%.
- AAPL: Exhibited a positive change of approximately 12.23%. It indicates a solid performance, though slightly lower than MSFT’s.
- GOOG: Showed a slight negative change of about -1.69%. It indicates a minor decline in its stock price over the observed period.
- NFLX: Experienced the most significant negative change, at approximately -11.07%. It suggests a notable decrease in its stock price during the period.

### 6. Risk vs Return Analysis.
- To calculate the average daily return and the standard deviation of daily returns for each stock. The standard deviation will serve as a proxy for risk, while the average daily return represents the expected return.

In [47]:
daily_returns = pivot_data.pct_change().dropna() 

In [50]:
# recalculating average daily return and standard deviation (risk)
avg_daily_return = daily_returns.mean()
risk = daily_returns.std()

In [58]:
# creating a DataFrame for plotting
risk_return_df = pd.DataFrame({'Risk': risk, 'Average Daily Return': avg_daily_return})

fig = go.Figure()

In [60]:
# adding scatter plot points

fig.add_trace(go.Scatter(
    x = risk_return_df['Risk'],
    y = risk_return_df['Average Daily Return'],
    mode = 'markers+text',
    text = risk_return_df.index,
    textposition = "top center",
    marker = dict(size=10)
))

# Updating layout
fig.update_layout(
    title = 'Risk vs. Return Analysis',
    xaxis_title = 'Risk (Standard Deviation)',
    yaxis_title = 'Average Daily Return',
    showlegend = False
)

fig.show()

###### Insight from the Risk vs Return Analysis:

- AAPL shows the lowest risk combined with a positive average daily return, suggesting a more stable investment with consistent returns.
- GOOG has higher volatility than AAPL and, on average, a slightly negative daily return, indicating a riskier and less rewarding investment during this period.
- MSFT shows moderate risk with the highest average daily return, suggesting a potentially more rewarding investment, although with higher volatility compared to AAPL.
- NFLX exhibits the highest risk and a negative average daily return, indicating it was the most volatile and least rewarding investment among these stocks over the analyzed period.